File: src/NTIX.Core/Models/DiffResult.cs
namespace NTIX.Core.Models;

public record DiffResult(
    List<PackageSpec> ToInstall = default!,
    List<PackageSpec> ToUpgrade = default!,
    List<PackageSpec> ToSkip = default!,
    List<PackageSpec> ToRemove = default!,
    List<PackageSpec> ToAdopt = default!,
    List<ScoopBucket> BucketsToAdd = default!,
    List<ScoopBucket> BucketsToRemove = default!,
    string? Error = null,
    List<string>? Warnings = null
)
{
    public List<PackageSpec> ToInstall { get; init; } = ToInstall ?? new();
    public List<PackageSpec> ToUpgrade { get; init; } = ToUpgrade ?? new();
    public List<PackageSpec> ToSkip { get; init; } = ToSkip ?? new();
    public List<PackageSpec> ToRemove { get; init; } = ToRemove ?? new();
    public List<PackageSpec> ToAdopt { get; init; } = ToAdopt ?? new();
    public List<ScoopBucket> BucketsToAdd { get; init; } = BucketsToAdd ?? new();
    public List<ScoopBucket> BucketsToRemove { get; init; } = BucketsToRemove ?? new();
    public string? Error { get; init; } = Error;
    public List<string> Warnings { get; init; } = Warnings ?? new();

    public bool IsEmpty => ToInstall.Count == 0 && ToUpgrade.Count == 0 && ToRemove.Count == 0 && ToAdopt.Count == 0
        && BucketsToAdd.Count == 0 && BucketsToRemove.Count == 0;
    public bool HasError => !string.IsNullOrEmpty(Error);
}

File: src/models/diff_result.rs
use crate::models::{
    config_file::ConfigFileEntry, manager_validation::ValidationResult, options::ScoopBucket,
    package_spec::PackageSpec,
};

#[derive(Debug, Clone, Default)]
pub struct DiffResult {
    pub to_install: Vec<PackageSpec>,
    pub to_upgrade: Vec<PackageSpec>,
    pub to_skip: Vec<PackageSpec>,
    pub to_remove: Vec<PackageSpec>,
    pub to_adopt: Vec<PackageSpec>,
    /// Packages installed on the system but not tracked in state. Informational
    /// only (not an action): run with `-a`/`--adopt` to adopt them.
    pub to_untracked: Vec<PackageSpec>,
    pub buckets_to_add: Vec<ScoopBucket>,
    pub buckets_to_remove: Vec<ScoopBucket>,
    pub config_files_to_create: Vec<ConfigFileEntry>,
    pub config_files_to_update: Vec<ConfigFileEntry>,
    /// Dest paths still tracked in state but no longer present in the config.
    pub config_files_no_longer_managed: Vec<String>,
    pub warnings: Vec<String>,
    /// Result of the single capability-detection pass for apply/diff runs.
    pub manager_validation: ValidationResult,
}

impl DiffResult {
    pub fn is_empty(&self) -> bool {
        self.to_install.is_empty()
            && self.to_upgrade.is_empty()
            && self.to_remove.is_empty()
            && self.to_adopt.is_empty()
            && self.buckets_to_add.is_empty()
            && self.buckets_to_remove.is_empty()
            && self.config_files_to_create.is_empty()
            && self.config_files_to_update.is_empty()
            && self.config_files_no_longer_managed.is_empty()
    }
}
